scripts/man/xmake.1
.TH "xmake" "1"
.SH NAME
xmake \- cross-platform build utility based on Lua


.SH SYNOPSIS
.B xmake
.RI [ task "] [" options "] [" target ]


.SH DESCRIPTION
.B xmake
is a lightweight cross-platform build utility based on Lua. It uses
.I xmake.lua
to maintain project builds. Compared with
.IR makefile / CMakeLists.txt ,
the configuration syntax is more concise and intuitive. It is very friendly to
novices and can quickly get started in a short time. Let users focus more on
actual project development.


.SH ACTIONS

.TP
.B b, build
Build targets if no given tasks.

.TP
.B u, uninstall
Uninstall the project binary files.

.TP
.B p, package
Package target.

.TP
.B r, run
Run the project target.

.TP
.B g, global
Configure the global options for xmake.

.TP
.B i, install
Package and install the target binary files.

.TP
.B c, clean
Remove all binary and temporary files.

.TP
.B create
Create a new project.

.TP
.B q, require
Install and update required packages.

.TP
.B update
Update and uninstall the xmake program.

.TP
.B f, config
Configure the project.


.SH PLUGINS

.TP
.B plugin
Manage plugins of xmake.

.TP
.B m, macro
Run the given macro.

.TP
.B doxygen
Generate the doxygen document.

.TP
.B l, lua
Run the lua script.

.TP
.B repo
Manage package repositories.

.TP
.B service
Start service for remote or distributed compilation and etc. (Experimental, still in development)

.TP
.B project
Generate the project file.

.TP
.B show
Show the given project information.


.SH OPTIONS

.TP
.B \-q, \-\-quiet
Quiet operation.

.TP
.B \-y, \-\-yes
Input yes by default if need user confirm.

.TP
.BR \-\-confirm =\fICONFIRM
Input the given result if need user confirm.
  \- yes
  \- no
  \- def

.TP
.B \-v, \-\-verbose
Print lots of verbose information for users.

.TP
.B \-\-root
Allow one to run xmake as root.

.TP
.B \-D, \-\-diagnosis
Print lots of diagnosis information (backtrace, check info ..) only for developers.
And we can append \fB\-v\fR to get more whole information.
  e.g. $ xmake \-vD

.TP
.B \-\-version
Print the version number and exit.

.TP
.B \-h, \-\-help
Print this help message and exit.

.TP
.BI \-F " FILE" ", \-\-file\fR=" FILE
Read a given
.B xmake.lua
file.

.TP
.BI \-P " PROJECT" ", \-\-project\fR=" PROJECT
Change to the given project directory.
Search priority:
  1. The Given Command Argument
  2. The Environment Variable: \fBXMAKE_PROJECT_DIR\fR
  3. The Current Directory


.SH BUILD OPTIONS

.TP
.B \-b, \-\-build
Build target. This is default building mode and optional.

.TP
.B \-r, \-\-rebuild
Rebuild the target.

.TP
.B \-a, \-\-all
Build all targets.

.TP
.B \-\-dry\-run
Dry run to build target.

.TP
.BI \-j " JOBS" ", \-\-jobs\fR=" JOBS
Specifies the number of jobs to build simultaneously. (default: 6)

.TP
.B \-w, \-\-warning
Enable the warnings output.

.TP
.BI \-\-files= FILES
Build the given source files.

e.g.
.RS
.EX
  \- xmake \-\-files=src/main.c
  \- xmake \-\-files='src/*.c' [target]
  \- xmake \-\-files='src/**c|excluded_file.c'
  \- xmake \-\-files='src/main.c:src/test.c'
.EE
.RE

.TP
.B target
The target name. It will build all default targets if this parameter is not specified.


.SH AUTHOR
.B xmake
is written by
.MT waruqi@\:gmail.com
ruki
.ME .

This manual page was written by
.MT mmyangfl@\:gmail.com
Yangfl
.ME
for the Debian Project (and may be used by others).


CHANGELOG.md
# Changelog ([中文](#中文))

## master (unreleased)

### New features

* [#4250](https://github.com/xmake-io/xmake/pull/4250): Improve link mechanism and order
* [#1438](https://github.com/xmake-io/xmake/issues/1438): Support code amalgamation
* [#3381](https://github.com/xmake-io/xmake/issues/3381): Add `xmake test` support
* [#4276](https://github.com/xmake-io/xmake/issues/4276): Support custom scope api
* [#4286](https://github.com/xmake-io/xmake/pull/4286): Add Apple XROS support

### Changes

* [#4284](https://github.com/xmake-io/xmake/issues/4284): Improve builtin includes

### Bugs fixed

* [#4256](https://github.com/xmake-io/xmake/issues/4256): Fix intellisense for vsxmake/c++modules

## v2.8.3

### New features

* [#4122](https://github.com/xmake-io/xmake/issues/4122): Support Lua Debugger (EmmyLua)
* [#4132](https://github.com/xmake-io/xmake/pull/4132): Support cppfront
* [#4147](https://github.com/xmake-io/xmake/issues/4147): Add hlsl2spv rule
* [#4226](https://github.com/xmake-io/xmake/issues/4226): Support sanitizers for package and policy
* Add lib.lua.package module
* Add `run.autobuild` policy
* Add global policies `xmake g --policies=`

### Changes

* [#4119](https://github.com/xmake-io/xmake/issues/4119): Improve to support emcc toolchain and emscripten package
* [#4154](https://github.com/xmake-io/xmake/issues/4154): Add `xmake -r --shallow target` to rebuild target without deps
* Add global ccache storage directory
* [#4137](https://github.com/xmake-io/xmake/issues/4137): Support Qt6 for Wasm
* [#4173](https://github.com/xmake-io/xmake/issues/4173): Add recheck argument to on_config
* [#4200](https://github.com/xmake-io/xmake/pull/4200): Improve remote build to support debugging xmake source code.
* [#4209](https://github.com/xmake-io/xmake/issues/4209): Add extra and pedantic warnings

### Bugs fixed

* [#4110](https://github.com/xmake-io/xmake/issues/4110): Fix extrafiles
* [#4115](https://github.com/xmake-io/xmake/issues/4115): Fix compile_commands generator for clangd
* [#4199](https://github.com/xmake-io/xmake/pull/4199): Fix compile_commands generator for c++ modules
* Fix os.mv fail on window
* [#4214](https://github.com/xmake-io/xmake/issues/4214): Fix rust workspace build error

## v2.8.2

### New features

* [#4002](https://github.com/xmake-io/xmake/issues/4002): Add soname and version support
* [#1613](https://github.com/xmake-io/xmake/issues/1613): Add avx512 and sse4.2 for add_vectorexts
* [#2471](https://github.com/xmake-io/xmake/issues/2471): Add set_encodings to set source/target encodings
* [#4071](https://github.com/xmake-io/xmake/pull/4071): Support the stm8 assembler on the sdcc toolchain.
* [#4101](https://github.com/xmake-io/xmake/issues/4101): Add force includes for c/c++
* [#2384](https://github.com/xmake-io/xmake/issues/2384): Add extrafiles for vs/vsxmake generator

### Changes

* [#3960](https://github.com/xmake-io/xmake/issues/3960): Improve msys2/crt64 support
* [#4032](https://github.com/xmake-io/xmake/pull/4032): Remove some old deprecated apis
* Improve to upgrade vcproj files in tools.msbuild
* Support add_requires("xmake::xxx") package
* [#4049](https://github.com/xmake-io/xmake/issues/4049): Improve rust to support cross-compilation
* Improve clang modules support

### Bugs fixed

* Fix exit all child processes on macOS/Linux

## v2.8.1

### New features

* [#3821](https://github.com/xmake-io/xmake/pull/3821): Add longpath option for windows installer
* [#3828](https://github.com/xmake-io/xmake/pull/3828): Add support for zypper package manager
* [#3871](https://github.com/xmake-io/xmake/issues/3871): Improve tools.msbuild to support for upgrading vsproj
* [#3148](https://github.com/xmake-io/xmake/issues/3148): Support grpc for protobuf
* [#3889](https://github.com/xmake-io/xmake/issues/3889): Support to add library path for add_links
* [#3912](https://github.com/orgs/xmake-io/issues/3912): Add set_pmxxheader to support objc precompiled header
* add_links support library file path

### Changes

* [#3752](https://github.com/xmake-io/xmake/issues/3752): Improve os.getenvs for windows
* [#3371](https://github.com/xmake-io/xmake/issues/3371): Improve tools.cmake to support ninja generator for wasm
* [#3777](https://github.com/xmake-io/xmake/issues/3777): Improve to find package from pkg-config
* [#3815](https://github.com/xmake-io/xmake/pull/3815): Improve tools.xmake to pass toolchains for windows
* [#3857](https://github.com/xmake-io/xmake/issues/3857): Improve to generate compile_commands.json
* [#3892](https://github.com/xmake-io/xmake/issues/3892): Improve to search packages from description
* [#3916](https://github.com/xmake-io/xmake/issues/3916): Improve to build swift program, support for multiple modules
* Update lua runtime to 5.4.6

### Bugs fixed

* [#3755](https://github.com/xmake-io/xmake/pull/3755): Fix find_tool from xmake/packages
* [#3787](https://github.com/xmake-io/xmake/issues/3787): Fix packages from conan 2.x
* [#3839](https://github.com/orgs/xmake-io/discussions/3839): Fix vs_runtime for conan 2.x

## v2.7.9

### New features

* [#3613](https://github.com/xmake-io/xmake/issues/3613): Add `wasm.preloadfiles` configuration for wasm
* [#3703](https://github.com/xmake-io/xmake/pull/3703): Support for conan >=2.0.5

### Changes

* [#3669](https://github.com/xmake-io/xmake/issues/3669): Improve cmake generator to support add_cxflags with the given tool
* [#3679](https://github.com/xmake-io/xmake/issues/3679): Improve `xrepo clean`
* [#3662](https://github.com/xmake-io/xmake/issues/3662): Improve cmake/make generator for lex/yacc project
* [#3697](https://github.com/xmake-io/xmake/issues/3662): Improve trybuild/cmake
* [#3730](https://github.com/xmake-io/xmake/issues/3730): Improve c++modules package installation

### Bugs fixed

* [#3596](https://github.com/xmake-io/xmake/issues/3596): Fix check_cxxfuncs and check_cxxsnippets
* [#3603](https://github.com/xmake-io/xmake/issues/3603): Fix `xmake update`

core/xmake.lua
-- project
set_project("xmake")

-- version
set_version("2.8.3", {build = "%Y%m%d"})

-- set xmake min version
set_xmakever("2.2.3")

-- set all warnings as errors
set_warnings("all", "error")

-- set language: c99, c++11
set_languages("c99", "cxx11")

-- add release and debug modes
add_rules("mode.release", "mode.debug")
if is_mode("release") then
    set_optimize("smallest")
    if is_plat("windows") then
        add_ldflags("/LTCG")
    end
end

-- disable some compiler errors
add_cxflags("-Wno-error=deprecated-declarations", "-fno-strict-aliasing", "-Wno-error=nullability-completeness", "-Wno-error=parentheses-equality")

-- add definitions
add_defines("_GNU_SOURCE=1", "_FILE_OFFSET_BITS=64", "_LARGEFILE_SOURCE")

-- add vectorexts
--[[
if is_arch("x86", "x64", "i386", "x86_64") then
    add_vectorexts("sse", "sse2", "sse3", "avx", "avx2")
elseif is_arch("arm.*") then
    add_vectorexts("neon")
end]]

-- for the windows platform (msvc)
if is_plat("windows") then
    set_runtimes("MT")
    add_ldflags("-nodefaultlib:msvcrt.lib")
    add_links("kernel32", "user32", "gdi32")
end

-- for mode coverage
if is_mode("coverage") then
    add_ldflags("-coverage", "-fprofile-arcs", "-ftest-coverage")
end

-- the runtime option
option("runtime")
    set_default("lua")
    set_description("Use luajit or lua runtime")
    set_values("luajit", "lua")
option_end()

-- the lua-cjson option
option("lua_cjson")
    set_default(true)
    set_description("Use lua-cjson as json parser")
option_end()

-- the readline option
option("readline")
    set_description("Enable or disable readline library")
    add_links("readline")
    add_cincludes("readline/readline.h")
    add_cfuncs("readline")
    add_defines("XM_CONFIG_API_HAVE_READLINE")
option_end()

-- the curses option
option("curses")
    set_description("Enable or disable curses library")
    add_links("curses")
    add_cincludes("curses.h")
    add_defines("XM_CONFIG_API_HAVE_CURSES")
option_end()

-- the pdcurses option
option("pdcurses")
    set_default(true)
    set_description("Enable or disable pdcurses library")
    add_defines("PDCURSES")
    add_defines("XM_CONFIG_API_HAVE_CURSES")
option_end()

-- only build xmake libraries for development?
option("onlylib")
    set_default(false)
    set_description("Only build xmake libraries for development")
option_end()

-- suppress warnings
if is_plat("windows") then
    add_defines("_CRT_SECURE_NO_WARNINGS")
    add_cxflags("/utf-8")
end

-- add projects
includes("src/sv", "src/lz4", "src/tbox", "src/xmake", "src/demo")
if has_config("lua_cjson") then
    includes("src/lua-cjson")
end
if is_config("runtime", "luajit") then
    includes("src/luajit")
else
    includes("src/lua")
end
if is_plat("windows") then
    includes("src/pdcurses")
end
